cpg_infra/abstraction/google_group_membership.py

```diff
@@ -0,0 +1,207 @@
+"""
+Contains pulumi.dynamic.ResourceProvider implementations for Google Groups Memberships
+"""
+
+from functools import cache
+from typing import TYPE_CHECKING, Any, Optional, TypedDict
+from urllib.parse import urlencode
+
+import googleapiclient.discovery
+import pulumi.dynamic
+from pulumi import Input, ResourceOptions
+
+# These don't work at runtime so only include for typing purposes
+if TYPE_CHECKING:
+    from googleapiclient._apis.cloudidentity.v1 import (  # type: ignore
+        CloudIdentityResource,
+    )
+
+MEMBER_LIST_PAGE_SIZE = 100
+
+
+class GroupMember(TypedDict):
+
+    member_key: str
+    "The email of the group member"
+
+    member_name: str
+    """
+    a path in the format groups/<group_id>/memberships/<member_id
+    where member_id is the gcloud id of the member, not the email.
+    This is needed when deleting memberships
+    """
+
+
+class GoogleGroupMembershipInputs(object):
+    group_id: Input[str]
+    member_key: Input[str]
+
+    def __init__(self, group_id: str, member_key: str):
+        self.group_id = group_id
+        self.member_key = member_key
+
+
+class GoogleGroupMembershipProviderInputs(TypedDict):
+    group_id: str
+    member_key: str
+
+
+class GoogleGroupMembership(pulumi.dynamic.Resource):
+    """A Pulumi dynamic resource for Google Group memberships."""
+
+    def __init__(
+        self,
+        name: str,
+        props: GoogleGroupMembershipInputs,
+        opts: Optional[ResourceOptions] = None,
+    ) -> None:
+        super().__init__(
+            GoogleGroupMembershipProvider(),
+            name,
+            {**vars(props)},
+            opts,
+        )
+
+
+class GoogleGroupMembershipProvider(pulumi.dynamic.ResourceProvider):
+    """A Pulumi dynamic resource provider for Google Groups settings."""
+
+    def create(self, props: GoogleGroupMembershipProviderInputs):
+        group_id = props['group_id']
+        member_key = props['member_key']
+
+        member = find_member_in_group_by_key(group_id, member_key)
+
+        # If the group membership already exists, then don't try and create it
+        if member != None:
+            return pulumi.dynamic.CreateResult(id_=member['member_name'])
+
+        # Otherwise create it here
+        created_member = add_member_to_group(group_id, member_key)
+
+        return pulumi.dynamic.CreateResult(id_=created_member['member_name'])
+
+    def delete(self, id: str, props: Any):
+        name_parts = id.split('/')
+        group_id = name_parts[1]
+        member = find_member_in_group_by_name(group_id, id)
+
+        # If the member has already been removed then no need to remove
+        if member == None:
+            return
+
+        remove_member_from_group(id)
+
+    def diff(self, id: str, olds: Any, news: Any):
+        # There isn't anything to change on a group membership, they are either a member
+        # or not, so this can always return false
+        return pulumi.dynamic.DiffResult(changes=False)
```

Comment:
I'm pretty sure, if you throw different params but the same key, I think pulumi will run this `diff`. And a different member_key should always cause a recreate with `replaces`: https://www.pulumi.com/docs/concepts/resources/dynamic-providers/#diffid-olds-news